### PR TITLE
#2377 MockOutputs validation before merge

### LINKED ISSUE #2377: SIGSEGV: segmentation violation code
terragrunt init command failed with the following message:

```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xedfd2e]


goroutine 96 [running]:

github.com/gruntwork-io/terragrunt/config.getTerragruntOutputIfAppliedElseConfiguredDefault({{0xc000c08a60, 0x8}, {0xc001747440, 0x15}, 0x0, 0x0, 0x0, 0x0, 0xc00033d860, 0x0}, ...)

	/home/circleci/project/config/dependency.go:343 +0x26e

github.com/gruntwork-io/terragrunt/config.(*Dependency).setRenderedOutputs(0xc00010d450, 0x1)

	/home/circleci/project/config/dependency.go:127 +0x11b

github.com/gruntwork-io/terragrunt/config.dependencyBlocksToCtyValue.func1()

	/home/circleci/project/config/dependency.go:285 +0x9c

golang.org/x/sync/errgroup.(*Group).Go.func1()

	/home/circleci/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67

created by golang.org/x/sync/errgroup.(*Group).Go

	/home/circleci/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
```

The code was the following:
```
include "state" {
  path = find_in_parent_folders()
}

include "drone" {
  path = find_in_parent_folders("drone.hcl")
}

dependency "postgres" {
  config_path = "../../postgres/common"

  mock_outputs_merge_strategy_with_state = "deep_map_only"
}

inputs = {
  postgres_user     = "recsys"
  postgres_password = dependency.postgres.outputs.user_passwords.recsys
}
```

### PR DESCRIPTION
**Pull request by denis256**
## Description

Fixes #2377.

Update logic to validate value of MockOutputs before execution of merge with state.

Included changes:
 * MockOutputs validation
 * Regression tests
 * `t.Helper()` execution in `logBufferContentsLineByLine` to print in logs source line from where output is coming.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

Updated evaluation of `mock_outputs_merge_strategy_with_state` in the `dependency` block to avoid crashes on empty mocks.

### CODE REVIEW

**Review by alikhil — APPROVED** [review]

**Review by Etiene — APPROVED** [review]
Great stuff!

File: config/dependency.go [hunk 1]
```diff
@@ -338,8 +338,14 @@ func getTerragruntOutputIfAppliedElseConfiguredDefault(dependencyConfig Dependen
 			case NoMerge:
 				return outputVal, nil
 			case ShallowMerge:
+				if dependencyConfig.MockOutputs == nil {
+					return outputVal, nil
+				}
```
Comment (brikis98) [inline config/dependency.go:343]: NIT: Should this check be above, around line 335? So we only enter this if statement block and switch statement if the `MockOutputs` isn't `nil`?

**Review by brikis98 — APPROVED** [review]

File: config/dependency.go (hunk 1 above)
Comment (denis256) [reply config/dependency.go]: Yes, simplified verification

**Review by brikis98 — APPROVED** [review]
LGTM

**Pull request merged**